### PR TITLE
Support PSUControl plugin

### PR DESCRIPTION
**Pull request by chriswal**
Automatically turn on printer with PSUControl plugin ( my fork has autoconnect). wait 20seconds for Printer to be up. then send the data for print.
Checkbox in config to turn on.

### CODE REVIEW

File: DiscoverOctoPrintAction.qml [hunk 1]
```diff
@@ -432,11 +432,23 @@ Cura.MachineAction
                         }
                     }
                     CheckBox
+                                        {
```
Comment (fieldOfView) [inline DiscoverOctoPrintAction.qml]: Please fix the indentation

File: DiscoverOctoPrintAction.qml [hunk 2]
```diff
@@ -432,11 +432,23 @@ Cura.MachineAction
...
+                        }
+                    }
+                    
+                     CheckBox
                     {
                         id: showCameraCheckBox
                         text: catalog.i18nc("@label", "Show webcam image")
                         enabled: manager.instanceSupportsCamera
-                        checked: manager.instanceApiKeyAccepted && Cura.ContainerManager.getContainerMetaDataEntry(Cura.MachineManager.activeMachineId, "octoprint_show_camera") != "false"
+                        checked: manager.instanceApiKeyAccepted && Cura.ContainerManager.getContainerMetaDataEntry(Cura.MachineManager.activeMachineId, "octoprint_show_camera") != "true"
```
Comment (fieldOfView) [inline DiscoverOctoPrintAction.qml]: This seems wrong/unintended.

File: OctoPrintOutputDevice.py [hunk 3]
```diff
@@ -367,7 +395,11 @@ def requestWrite(self, nodes: List["SceneNode"], file_name: Optional[str] = None
                 return
 
         self._startPrint()
-
+
```
Comment (fieldOfView) [inline OctoPrintOutputDevice.py]: These newlines seem extraneous. Please revert these.

File: DiscoverOctoPrintAction.qml (hunk 1 above)
Comment (chriswal) [reply DiscoverOctoPrintAction.qml]: done

File: DiscoverOctoPrintAction.qml (hunk 2 above)
Comment (chriswal) [reply DiscoverOctoPrintAction.qml]: this is a oneliner ...looks unintended because of the new line ...it starts at checked:

**Comment by chriswal** [comment]
i can try if you can point me in the right direction please?
the plugins you have in a list _plugin.data ? discoverOctoprintAction.py
and the items need to be influenced is checkbox ..enabled and checked  ? discoverOctoprintAction.qml
i dont know how to connect these two

File: OctoPrintOutputDevice.py (hunk 3 above)
Comment (chriswal) [reply OctoPrintOutputDevice.py]: done

**Comment by chriswal** [comment]
setting is now disabled if PSUControl is not installed in the Octoprint instance

**Comment by fieldOfView** [comment]
I have some proposed style-changes here:
https://github.com/chriswal/OctoPrintPlugin/compare/3.5...fieldOfView:psu_control

* make DiscoverOctoPrintAction changes less specific to this plugin
* hide control if plugin is not installed (instead of disabling it)
* use existing _plugin_data instead of a new plugin-specific variable in OctoPrintOutputDevice
* remove unused API; checking and turning off were never used anyway

I am wondering if the 20 seconds delay is necessary. AFAIK, the controller is turned on and connected in order to receive the gcode to turn on the PSU. Why would that need such a long delay? I can imagine a delay is useful, but would rather give it a second or so.

Also, can you enlighten me why the "Trigger commands" on the OctoPrint end aren't good enough?

**Comment by fieldOfView** [comment]
I see now, both the delay and the Trigger commands not being a suitable solution are explained by using eg a relay and a system command instead of a gcode command with the PSU Control plugin.

**Comment by chriswal** [comment]
I dont really know what i am doing here so any input is welcome ;-) i tried with looking at your code, and copying it, to get the output i want.
yes its a great idea to make it less plugin specific ...so it is easy to add different plugins ..
then also hiding is better.

yes my printer takes 6 seconds to boot after that octoprint needs to connect. thats why i changed the psucotrol plugin to connect automatically after power on ..because if you use another plugin for discovery that would add time. i used 15 seconds for my printer and it worked fine. but with the new istalled octoprint 1.3.11 it didnt work one time so to be sure i added 5 seconds.

perhaps i should add the timeout to the settings?
and also a check if the plugin is disabled before sending the power on command?

**Comment by fieldOfView** [comment]
For inspiration, could you add a dump the plugins section of your /api/settings?apikey=[key] ?

**Comment by chriswal** [comment]
you meant the octoprint config?
or the api of psucontrol?

**Comment by chriswal** [comment]
ok i see ... one moment

**Comment by fieldOfView** [comment]
I meant the output from http://[octoprint address]/api/settings?apikey=[your api key]

**Comment by chriswal** [comment]
i always get the full config so i cut out the psucontrol part.

```jason
"psucontrol": {
      "GPIOMode": "BOARD", 
      "autoConnect": true, 
      "autoConnectBaudrate": "", 
      "autoConnectOnPowerON": true, 
      "autoConnectPort": "/dev/serial/by-id/usb-1a86_USB2.0-Serial-if00-port0", 
      "autoConnectPrinterProfile": "", 
      "autoConnectWaitTimeout": "6.1", 
      "autoOn": false, 
      "autoOnTriggerGCodeCommands": "G0,G1,G2,G3,G10,G11,G28,G29,G32,M104,M106,M109,M140,M190", 
      "disconnectOnPowerOff": true, 
      "enablePowerOffWarningDialog": false, 
      "enablePseudoOnOff": true, 
      "idleIgnoreCommands": "M105", 
      "idleTimeout": "20", 
      "idleTimeoutWaitTemp": 50, 
      "invertonoffGPIOPin": false, 
      "invertsenseGPIOPin": false, 
      "offGCodeCommand": "M81", 
      "offSysCommand": "curl \"http://10.0.0.2/SWITCH.CGI?s1=0\"", 
      "onGCodeCommand": "M80", 
      "onSysCommand": "curl \"http://10.0.0.2/SWITCH.CGI?s1=1\"", 
      "onoffGPIOPin": 0, 
      "postOnDelay": "0", 
      "powerOffWhenIdle": true, 
      "pseudoOffGCodeCommand": "M81", 
      "pseudoOnGCodeCommand": "M80", 
      "senseGPIOPin": 0, 
      "senseGPIOPinPUD": "", 
      "senseSystemCommand": "", 
      "sensingMethod": "INTERNAL", 
      "switchingMethod": "SYSTEM"
    }, `

**Comment by fieldOfView** [comment]
I'll merge the PR with my changes, and we can improve from there.

**Pull request merged**